relayer/shannon/fullnode_cache.go

```diff
@@ -0,0 +1,154 @@
+	appCacheMu sync.RWMutex
+
+	// TODO_IMPROVE: Add a sessionCacheKey type with the necessary helpers to concat a key
+	// sessionCache caches sessions for use by the Relay function.
+	// map keys are of the format "serviceID-appID"
+	sessionCache   map[string]sessiontypes.Session
+	sessionCacheMu sync.RWMutex
+}
+
+// start launches a goroutine, only once per instance of FullNodeCache, to
+func (cfn *CachingFullNode) start() {
+	go func() {
+		// TODO_IMPROVE: make the refresh interval configurable.
+		ticker := time.NewTicker(time.Minute)
+		for {
+			cfn.updateAppCache()
+			cfn.updateSessionCache()
+
+			<-ticker.C
+		}
+	}()
+}
+
+func (cfn *CachingFullNode) GetServiceApps(serviceID relayer.ServiceID) ([]apptypes.Application, error) {
+	cfn.appCacheMu.RLock()
+	defer cfn.appCacheMu.RUnlock()
+
+	apps, found := cfn.appCache[serviceID]
+	if !found {
+		return nil, fmt.Errorf("getServiceApps: no apps found for service %s", serviceID)
+	}
+
+	return apps, nil
+}
+
+func (cfn *CachingFullNode) GetSession(serviceID relayer.ServiceID, appAddr string) (sessiontypes.Session, error) {
+	cfn.sessionCacheMu.RLock()
+	defer cfn.sessionCacheMu.RUnlock()
+
+	session, found := cfn.sessionCache[sessionCacheKey(serviceID, appAddr)]
+	if !found {
+		return session, fmt.Errorf("getSession: no cached sessions found for service %s, app %s", serviceID, appAddr)
+	}
+
+	return session, nil
+}
+
+// SendRelay delegates the sending of the relay to the LazyFullNode.
+func (cfn *CachingFullNode) SendRelay(app apptypes.Application, session sessiontypes.Session, endpoint endpoint, payload relayer.Payload) (*servicetypes.RelayResponse, error) {
+	return cfn.LazyFullNode.SendRelay(app, session, endpoint, payload)
+}
+
+func (cfn *CachingFullNode) IsHealthy() bool {
+	cfn.appCacheMu.RLock()
+	defer cfn.appCacheMu.RUnlock()
+	cfn.sessionCacheMu.RLock()
+	defer cfn.sessionCacheMu.RUnlock()
+
+	return len(cfn.appCache) > 0 && len(cfn.sessionCache) > 0
+}
+
+func (cfn *CachingFullNode) updateAppCache() {
+	appData, err := cfn.LazyFullNode.GetAllServicesApps()
+	if err != nil {
+		cfn.Logger.Warn().Err(err).Msg("updateAppCache: error getting the list of apps; skipping update.")
+		return
+	}
+
+	cfn.appCacheMu.Lock()
+	defer cfn.appCacheMu.Unlock()
+	cfn.appCache = appData
+}
+
+func (cfn *CachingFullNode) updateSessionCache() {
+	sessions := cfn.fetchSessions()
+	if len(sessions) == 0 {
+		cfn.Logger.Warn().Msg("updateSessionCache: received empty session list; skipping update.")
+		return
+	}
+
+	cfn.sessionCacheMu.Lock()
+	defer cfn.sessionCacheMu.Unlock()
+	cfn.sessionCache = sessions
+}
+
+func (cfn *CachingFullNode) fetchSessions() map[string]sessiontypes.Session {
+	logger := cfn.Logger.With("method", "fetchSessions")
+
+	apps, err := cfn.LazyFullNode.GetAllServicesApps()
+	if err != nil {
+		logger.Warn().Err(err).Msg("fetchSession: error listing applications")
+	}
+
+	sessions := make(map[string]sessiontypes.Session)
+	// TODO_TECHDEBT: use multiple go routines.
```

Comment:
Do you really think it's necessary?

Reply:
Yes, for two reasons:
1. Shannon allows a larger number of applications.
2. A fullnode could have a hiccup on a one or more requests, delaying the entire relay serving functionality.
3. This is a critical section of the code, which needs to complete as soon as possible to allow a PATH instance to start serving requests.